sno/working_copy/sqlserver.py

```diff
@@ -0,0 +1,553 @@
+
+        return new_type == old_type
+
+    def _remove_hidden_meta_diffs(self, dataset, ds_meta_items, wc_meta_items):
+        super()._remove_hidden_meta_diffs(dataset, ds_meta_items, wc_meta_items)
+
+        # Nowhere to put these in postgis WC
+        for key in self._UNSUPPORTED_META_ITEMS:
+            if key in ds_meta_items:
+                del ds_meta_items[key]
+
+        # Diffing CRS is not yet supported.
+        for key in list(ds_meta_items.keys()):
+            if key.startswith("crs/"):
+                del ds_meta_items[key]
+
+    def _is_meta_update_supported(self, dataset_version, meta_diff):
+        """
+        Returns True if the given meta-diff is supported *without* dropping and rewriting the table.
+        (Any meta change is supported if we drop and rewrite the table, but of course it is less efficient).
+        meta_diff - DeltaDiff object containing the meta changes.
+        """
+        # For now, just always drop and rewrite.
+        return not meta_diff
+
+
+class InstanceFunction(Function):
+    """
+    An instance function that compiles like this when applied to an element:
+    >>> element.function()
+    Unlike a normal sqlalchemy function which would compile as follows:
+    >>> function(element)
+    """
+
+    pass
+
+
+@compiles(InstanceFunction)
+def compile_instance_function(element, compiler, **kw):
+    return "(%s).%s()" % (element.clauses, element.name)
+
+
+class GeographyType(UserDefinedType):
+    """UserDefinedType so that V2 geometry is adapted to MS binary format."""
+
+    def __init__(self, crs_id):
+        self.crs_id = crs_id
+
+    def bind_processor(self, dialect):
+        # 1. Writing - Python layer - convert sno geometry to WKB
+        return lambda geom: geom.to_wkb()
+
+    def bind_expression(self, bindvalue):
+        # 2. Writing - SQL layer - wrap in call to STGeomFromWKB to convert WKB to MS binary.
+        return Function(
+            quoted_name("geography::STGeomFromWKB", False),
+            bindvalue,
+            self.crs_id,
+            type_=self,
+        )
+
+    def column_expression(self, col):
+        # 3. Reading - SQL layer - append with call to .STAsBinary() to convert MS binary to WKB.
+        return InstanceFunction("STAsBinary", col, type_=self)
+
+    def result_processor(self, dialect, coltype):
+        # 4. Reading - Python layer - convert WKB to sno geometry.
+        return lambda wkb: Geometry.from_wkb(wkb)
+
+
+class BaseDateOrTimeType(UserDefinedType):
+    """UserDefinedType so we read dates, times, and datetimes as text."""
+
+    def column_expression(self, col):
+        # When reading, convert dates and times to strings using style 127: ISO8601 with time zone Z.
+        return Function(
+            "CONVERT",
+            literal_column("NVARCHAR"),
+            col,
+            literal_column("127"),
+            type_=self,
+        )
+
+
+def sqlserver_upsert(*args, **kwargs):
```

Comment:
This is generic to MSSQL but not our working copy? Would it be better homed in `sqlserver_adapter.py`?

Reply:
I'm open to moving everything around (maybe not in this PR).

Right now sqlserver_adapter (and other adapter classes) deal specifically with how the Sno model maps onto the information we can store inside the working copy of that type. And then sqlserver.py has basically everything else needed for this WC. So, current location makes sense from that logic, but open to other ways of dividing things up.

I might move it to an upsert file that specifies how to do upserts for all 3 dialects - that will get it out of the way.

Reply:
> Right now sqlserver_adapter (and other adapter classes) deal specifically with how the Sno model maps onto the information we can store inside the working copy of that type. 

Ah, that was my misunderstanding.

My though process was that:
- upsert/function/type-handling code is stuff that might appear in SQLAlchemy one day (or we could contribute)
- if we want native SQL _importers_ in future we'd likely need it, separately to use in our working copies.

I don't feel strongly about it though 😄